display "(empty range)" in case of empty ranges

prashantrahul141 opened the pull request:
As mentioned in #40331, empty ranges could be confusing. This pr adds a message "(empty range)" incase if the range is empty. 
So for the example given in the issue
```sh
julia> a = 1:3
1:3

julia> b = 5:6
5:6

julia> intersect(a,b)  # old behaviour
5:4

julia> intersect(a,b)  # new behaviour
5:4 (empty range)
```

Seelengrab commented on base/range.jl:
```diff
@@ -618,6 +618,9 @@ function show(io::IO, r::LinRange{T}) where {T}
     print(io, ", ")
     show(io, length(r))
     print(io, ')')
+    if isempty(r)
```
This is the 2-argument `show`, which should be parseable Julia code. See https://docs.julialang.org/en/v1/base/io-network/#Base.show-Tuple{IO,%20Any} .

prashantrahul141 replied:
would it be better to use `compact` in `IOContext` to control this like @mikmoore suggested [here](https://github.com/JuliaLang/julia/pull/57697#issuecomment-2710893642)?

basically this
```julia
    if !get(io, :compact, false) && isempty(r)
        print(io, " (empty range)")
    end
```

KristofferC replied:
I think it should be in a three arg show method.

Seelengrab replied:
Yes, that's what I was getting at. From the docstring I linked:

> For a more verbose human-readable text output for objects of type ` T`, define `show(io::IO, ::MIME"text/plain", ::T)` in addition.

` :compact` showing is unrelated to this.

prashantrahul141 replied:
I see now. So I assume it would be better to define `show(io::IO, ::MIME"text/plain", ::T)` for range types rather than change the existing ones?

mikmoore commented:
Can we tweak this so that it doesn't print when IOContext is `compact`?

And also try to minimize its interference with the ability to copy-paste to reconstruct an object (which might be addressed by the `compact`)? For example,
```julia-repl
julia> print([1:2, 3:2, 4:5])
UnitRange{Int64}[1:2, 3:2, 4:5]
```
I can use `UnitRange{Int64}[1:2, 3:2, 4:5]` to reconstruct this object but `UnitRange{Int64}[1:2, 3:2 (empty range), 4:5]` is a `ParseError`.

jishnub commented:
If I understand correctly, the issue is mainly with an empty `UnitRange` being confused with a `StepRange` with a step of `-1`? Do we need this for every range type?

prashantrahul141 commented:
I assume a better approach would be to define a 3-argument show (`show(io::IO, ::MIME"text/plain", ::T)`) for range types. As @jishnub mentioned, should it be defined for all range types or just for `UnitRange`?

I think it should be defined for all range types since 3-argument show is meant to display additional details and to ensure consistency.

jishnub commented:
To take the `LinRange` example, here are the displayed forms for the 3 and 2-argument `show` methods:
```julia
julia> show(stdout, MIME"text/plain"(), LinRange(2.0, 4.5, 5))
5-element LinRange{Float64, Int64}:
 2.0, 2.625, 3.25, 3.875, 4.5

julia> show(stdout, LinRange(2.0, 4.5, 5))
LinRange{Float64}(2.0, 4.5, 5)
```
In the first case, the length is explicitly mentioned in the summary. In the second case, the last parameter corresponds to the length, although this isn't explicitly pointed out (as this needs to be runnable julia code). Since we're discussing the 3-argument case here, the displayed form seems to indicate the length adequately?

I feel the issue is mainly with ranges where the 3-argument `show` uses the 2-argument method for the actual display. Or, in other words, any range that does not specialize the 3-argument `show` method. `LinRange` and `LogRange` are ones where the 3-argument form does display the summary, so these may be excluded.

prashantrahul141 commented:
So I am trying to define a 3 argument `show` in `base/range.jl`, something like:
```julia
function Base.show(io::IO, ::MIME"text/plain", r::AbstractUnitRange{T}) where T
   # impl
end
```
and I am getting this
```sh
error during bootstrap:
LoadError(at "Base_compiler.jl" line 3: LoadError(at "range.jl" line 1119: LoadError(at "range.jl" line 1119: UndefVarError(var=:var"@MIME_str", world=0x0000000000000e4e, scope=Base))))
ijl_undefined_var_error at /home/prashant/Projects/julia-fork/src/rtutils.c:154
jl_eval_global_var at /home/prashant/Projects/julia-fork/src/interpreter.c:169 [inlined]
eval_value at /home/prashant/Projects/julia-fork/src/interpreter.c:223
jl_interpret_toplevel_expr_in at /home/prashant/Projects/julia-fork/src/interpreter.c:919
No code info - unknown interpreter state!
```
I am assuming this is because maybe `MIME` is not available at this stage of bootstraping. How can I get around this?

nsajko commented:
> not available at this stage of bootstraping. How can I get around this?

`MIME` and friends are includes into `Base` here:

https://github.com/JuliaLang/julia/blob/885b1cd875f101f227b345f681cc36879124d80d/base/Base.jl#L63-L64

... so try inserting your method somewhere after that point, maybe try one of these files:

https://github.com/JuliaLang/julia/blob/885b1cd875f101f227b345f681cc36879124d80d/base/Base.jl#L112-L114

prashantrahul141 committed:
impl: 3 argument show method for UnitRange

prashantrahul141 committed:
fix: doctest

LilithHafner reviewed (approved):
@mikmoore, this PR doesn't actually change the output in your example. When folks ask for parseable Julia code (2-arg show), containers recursively call 2-arg show on their elements so that the elements parse. This is behavior using this PR:

```julia-repl
julia> println([1:2, 3:2, 4:5])
UnitRange{Int64}[1:2, 3:2, 4:5]

julia> 3:2
3:2 (empty range)

julia> [1:2, 3:2, 4:5]
3-element Vector{UnitRange{Int64}}:
 1:2
 3:2 (empty range)
 4:5
```

---

This PR currently only applies to UnitRange, which is minimal and IMO sufficient to address the motivating confusion about `4:3` vs `4:-1:3`

---

I'll go ahead and merge unless anyone objects.

mcabbott commented:
Note that it also affects printing within a matrix, at present:
```julia
julia> [rand(1:3):rand(1:3) for _ in 1:4]
4-element Vector{UnitRange{Int64}}:
 2:1 (empty range)
 1:3
 1:3
 2:1 (empty range)

julia> [rand(1:3):rand(1:3) for _ in 1:4, _ in 1:5]
4×5 Matrix{UnitRange{Int64}}:
 2:1 (empty range)  3:2 (empty range)  2:1 (empty range)  3:2 (empty range)  3:3
 3:3  2:2  2:1 (empty range)  1:3  3:3
 2:3  3:3  1:1  3:3  1:1
 2:2  2:2  2:3  3:2 (empty range)  3:2 (empty range)
```

LilithHafner commented:
The Vector case looks good but the matrix case is annoying. Perhaps we should simply check the compact flag and skip the hint for all collections. (https://github.com/JuliaLang/julia/pull/57697#discussion_r1987830730)

prashantrahul141 committed:
check for :compact in IO before printing

from the docs:
"Checking the :compact IOContext key (often checked as get(io, :compact,
false)::Bool) of io in such methods is recommended, since some
containers show their elements by calling this method with :compact =>
true."

https://docs.julialang.org/en/v1/base/io-network/#Base.show-Tuple{IO,%20Any}

nsajko commented on base/show.jl:
```diff
@@ -12,6 +12,13 @@ end
 show(io::IO, ::MIME"text/plain", r::AbstractRange) = show(io, r) # always use the compact form for printing ranges
 
+function show(io::IO, ::MIME"text/plain", r::UnitRange)
+    print(io, repr(first(r)), ':', repr(last(r)))
```
More efficient:

```suggestion
    show(io, first(r))
    print(io, ':')
    show(io, last(r))
```

Proof:

```julia-repl
julia> f(io, a, b) = print(io, repr(a), ':', repr(b))
f (generic function with 1 method)

julia> function g(io, a, b)
           show(io, a)
           print(io, ':')
           show(io, b)
       end
g (generic function with 1 method)

julia> using BenchmarkTools

julia> @btime f(io, a, b) setup=(io = devnull; a = big(2)^(2^10); b = a + 1;)
  2.547 μs (12 allocations: 1.62 KiB)

julia> @btime g(io, a, b) setup=(io = devnull; a = big(2)^(2^10); b = a + 1;)
  2.201 μs (4 allocations: 736 bytes)
```

prashantrahul141 committed:
perf: use multi stage show, print

https://github.com/JuliaLang/julia/pull/57697#discussion_r2001497340

nsajko commented on base/show.jl:
```diff
@@ -12,6 +12,15 @@ end
+    show(io, first(r))
+    print(io, ':')
+    show(io, last(r))
+    if !get(io, :compact, false) && isempty(r)
```
I'd type assert `Bool` here just to be safe, because `typeof(io)` is often not known concretely at compilation time (`stdout` is a global variable that's not concretely typed). Might make some code less vulnerable to invalidation.

```suggestion
    if !(get(io, :compact, false)::Bool) && isempty(r)
```

prashantrahul141 commented:
😅 thank you!

prashantrahul141 committed:
type assert get() call

https://github.com/JuliaLang/julia/pull/57697#discussion_r2001579971

LilithHafner commented on base/show.jl:
```diff
@@ -12,6 +12,15 @@ end
 show(io::IO, ::MIME"text/plain", r::AbstractRange) = show(io, r) # always use the compact form for printing ranges
 
+function show(io::IO, ::MIME"text/plain", r::UnitRange)
+    show(io, first(r))
+    print(io, ':')
+    show(io, last(r))
```
```suggestion
    show(io, r)
```
The implementation here should match the 2-arg implementation (so might as well call the 2-arg method). 

While @nsako's suggestion is reasonable, it shouldn't only be applied to the 3-arg and not the 2-arg method because that would cause inconsistencies like this:

```julia-repl
julia> io = IOContext(stdout, :typeinfo=>Bool);

julia> show(io, MIME("text/plain"), false:true)
0:1
julia> show(io, MIME("text/plain"), false:true:true)
false:true:true
```

I would recommend attempting that performance improvement in a separate pull request due to its semantic impact. Calling `show(io, r)` here means that any performance improvements to `show(io, r)` will also apply here.

prashantrahul141 committed:
internally call 2 arg show to be consistent

the perf improvement should be moved to the 2 arg show implementation
instead

LilithHafner reviewed (approved):
@nsajko, is this okay with you?

prashantrahul141 commented:
I have also added that performance improvement to `show(io, r)` in another pr here #57823  as mentioned by @LilithHafner https://github.com/JuliaLang/julia/pull/57697#discussion_r2001723774

stevengj commented:
Can I make an alternative suggestion?  Define
```jl
function show(io::IO, ::MIME"text/plain", r::AbstractRange)
    summary(io, r)
    println()
    show(io, r)
end
```
which will print:
```
julia> 1:100
100-element UnitRange{Int64}
1:100

julia> 1:0
0-element UnitRange{Int64}
1:0

julia> 1:0.1:10
91-element StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}
1.0:0.1:10.0
```
This way it
* looks more like the text/plain display of other `AbstractArray` types.
* will still use the 2-argument `show` format when printing inside containers (due to our [annoying heuristic based on detection of a newline](https://github.com/JuliaLang/julia/pull/34387#issuecomment-2127778976))
